src/webcore/executor.rs

```diff
@@ -0,0 +1,226 @@
+        js! { @(no_return)
+            @{&self.waker}.drop();
+        }
+    }
+}
+
+impl EventLoopInner {
+    // Initializes the event loop. Only called once.
+    fn new() -> Self {
+        EventLoopInner {
+            microtask_queue: RefCell::new(VecDeque::with_capacity(INITIAL_QUEUE_CAPACITY)),
+            waker: js!(
+                var callback = @{|| EventLoop.drain()};
+                var wrapper = function() {
+                    if (!callback.dropped) { callback() }
+                };
+
+                // Modern browsers can use `MutationObserver` which allows
+                // us to schedule a micro-task without allocating a promise.
+                // https://dom.spec.whatwg.org/#notify-mutation-observers
+                if ( typeof MutationObserver === "function" ) {
+                    var node = document.createTextNode( "0" );
+                    var state = false;
+
+                    new MutationObserver( wrapper ).observe( node, { characterData: true } );
+
+                    function nextTick() {
+                        state = !state;
+                        node.data = ( state ? "1" : "0" );
+                    }
+
+                // Node.js and other environments
+                } else {
+                    var promise = Promise.resolve( null );
+
+                    function nextTick() {
+                        promise.then( wrapper );
+                    }
+                }
+
+                nextTick.drop = function() {
+                    callback.dropped = true;
+                    callback.drop();
+                };
+
+                return nextTick;
+            ).try_into().unwrap(),
+            shrink_counter: Cell::new(0)
+        }
+    }
+    // Pushes a task onto the queue
+    fn push_task(&self, task: Rc< SpawnedTask >) {
+        let mut queue = self.microtask_queue.borrow_mut();
+        queue.push_back(task);
+
+        // If the queue was previously empty, then we need to schedule
+        // the queue to be drained.
+        if queue.len() == 1 {
+            self.wake();
+        }
+    }
+    // Invoke the JavaScript waker function
+    fn wake(&self) {
+        js! { @(no_return) @{&self.waker}(); }
+    }
+    // Remove and return a task from the front of the queue
+    fn pop_task(&self) -> Option< Rc< SpawnedTask > > {
+        self.microtask_queue.borrow_mut().pop_front()
+    }
+    fn shrink_if_necessary(&self) {
+        let mut queue = self.microtask_queue.borrow_mut();
+        // We consider shrinking the queue if it is less than
+        // half full...
+        if queue.len() <= queue.capacity() / 2 {
+            // ...and if it's been that way for at least
+            // `QUEUE_SHRINK_DELAY` iterations.
+            let shrink_counter = self.shrink_counter.get();
+            if shrink_counter < QUEUE_SHRINK_DELAY {
+                self.shrink_counter.set(shrink_counter+1);
+            } else {
+                queue.shrink_to_fit();
```

Comment:
It shouldn't use `shrink_to_fit`, it should replace the queue with a new `VecDeque` with `INITIAL_QUEUE_CAPACITY`

Reply:
`shrink_to_fit` also deallocates + allocates, but less efficiently than recreating the `VecDeque`.

In addition, `shrink_to_fit` will cause it to have 0 capacity, but we want it to have `INITIAL_QUEUE_CAPACITY` capacity.

Reply:
No, we want it to have a capacity equal to the last queue size, since that is a good estimate for future iterations. We could avoid some overhead by waiting to resize it until after draining the items, but this is a very cold code-path (it's only hit 25 iterations after sudden drops in the queue size).

Reply:
Ahh, I see, you moved the call to `shrink_if_necessary`

I still don't like that it might have a capacity lower than `INITIAL_QUEUE_CAPACITY`

Reply:
There's actually not a good way to control the capacity this way. I've implemented the closest approximation instead, and opened https://github.com/rust-lang/rust/issues/49385.

Reply:
Sure there is... have the `shrink_if_necessary` method return an `Option<usize>`, and then after draining it will check if it is `Some`, and if so it will replace the old `VecDeque` with a new `VecDeque` with the appropriate capacity.

Reply:
Yeah I didn't really want to leak this complexity into the drain function, but I've implemented that now.